common/core/web/input-processor/src/text/inputProcessor.ts

```diff
@@ -76,7 +76,7 @@ namespace com.keyman.text {
 
       // Will handle keystroke-based non-layer change modifier & state keys, mapping them through the physical keyboard's version
       // of state management.
-      if(!fromOSK && this.keyboardProcessor.doModifierPress(keyEvent, outputTarget, !fromOSK)) {
+      if(this.keyboardProcessor.doModifierPress(keyEvent, outputTarget, !fromOSK) && !fromOSK) {
```

Comment:
Keep in mind that with boolean operator [short-circuit evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation), the `&&` is _**not**_ an associative operator - order matters.

We always want `doModifierPress` to run, as that triggers OSK state-key updates.  We simply don't want to instantly exit the function unless it's also caused by a physical keystroke.

Reply:
When we have side-effects like this, I like to structure the code to emphasize it:
```suggestion
      if(this.keyboardProcessor.doModifierPress(keyEvent, outputTarget, !fromOSK)) {
        if(!fromOSK) {
```

Makes it slightly less likely that a future maintainer won't accidentally swap them back...